pages-update

DmitriiTsy opened the pull request:
Created pages functionality (testing)

DmitriiTsy committed:
pages-update

DmitriiTsy committed:
api-error-fix

DmitriiTsy committed:
functionality-update

DmitriiTsy committed:
styling-updates

DmitriiTsy commented:
**Basic Description:** 
The first page of the Rick&Morty web site the functionality of:
1. Choosing the Characters name (Input search) 
2. Filtering Characters (there are categories of filters)

**Detailed description:** 
Every time when the user chooses the name/category of a character I'm creating a unique API request with this data

**Data used:** 
Used https://rickandmortyapi.com/ as an API for collecting data (status, species, gender of elements and etc)

anatoliy8493 commented on src/App.css:
```diff
@@ -1,16 +1,6 @@
-    animation: App-logo-spin infinite 20s linear;
-  }
+html, body {
+  width: 100%;
```
let's add normalize.css instead

DmitriiTsy replied:
Sure, I linked normalized CSS to the App.js file and deleted this custom thing from App.css

anatoliy8493 commented on src/App.css:
```diff
@@ -1,16 +1,6 @@
-  text-align: center;
-}
-
-.App-logo {
```
are these changes from the main already? should I review another PR first?

anatoliy8493 commented on src/App.js:
```diff
@@ -1,25 +1,28 @@
   return (
-      <div>
-         <img src={BackgroundImage} className="background-image"></img>
+      <div className='main-container'>
```
I'd recommend using double quotes for react string props

DmitriiTsy replied:
I will replace it everywhere. And will use only it in the future

anatoliy8493 commented on src/App.js:
```diff
@@ -1,25 +1,28 @@
-          <button onClick={HandlerCharacters} className='characters-button'>CHARACTERS</button>
-          <button className='characters-button'>LOCATIONS</button>
-          <button className='characters-button'>EPISODES</button>
+            <button onClick={HandlerCharacters} style={activeCharacters === true? {borderBottom: '2px solid orange'} : {border: 'none'}} className='characters-button'>CHARACTERS</button>
```
it's a bad practice to write inline styles ^, can we use [classnames](https://github.com/JedWatson/classnames) instead?

DmitriiTsy replied:
Sure, done

anatoliy8493 commented on src/components/Characters/Character.css:
```diff
@@ -3,16 +3,16 @@
 }
 
 .Character__image{
```
let's use kebab-case for css ^

DmitriiTsy replied:
sure, I believe now I have a "muscle memory" about using it

anatoliy8493 commented on src/components/Characters/Character.js:
```diff
@@ -2,27 +2,53 @@ import React, {useState} from 'react'
 
     const [TouchImage, setTouchImage] = useState(false)
 
     const HandlerTouch = (event) => {
```
why it's capitalized? should be lowerCamelCase

DmitriiTsy replied:
sure

anatoliy8493 commented on src/components/Characters/Character.js:
```diff
@@ -2,27 +2,53 @@ import React, {useState} from 'react'
 
 const Character = ({props, index}) => {
-    const {image,name,status,species,gender} = props
+    const {image,name,status,species,gender, episode} = props
```
bad formatting, I'd recommend installing ESLint and extend from the Airbnb to make this auto-fixable once you save changes

DmitriiTsy replied:
Sure

anatoliy8493 commented on src/components/Characters/Characters.js:
```diff
@@ -1,13 +1,32 @@
-    const [results, setResults] =useState([])
-    const [search, setSearch] =useState('')
+    const [url, setUrl] = useState('https://rickandmortyapi.com/api/character/')    // Fetching charact. API
+    const [info, setInfo] = useState({});             
```
too many useStates in one file, can we split the up?

DmitriiTsy replied:
Sure, done

anatoliy8493 commented on src/components/Characters/Characters.js:
```diff
@@ -17,38 +36,175 @@ const Characters = () => {
-  
+    }, [search, page, url, currentStatus, currentGender, currentSpecies, error])                       //Serach as a main dependency for this 
+
+    const HandlerBefore = (event) => {
```
HandlerBefore has 3 problems in the same name
 - should start from the lower case latter
 - the first word in the name should be a verb since function = action
 - before is too general to name a handler, any other name could be better, ie handleButtonClick or handleImagesLoad, etc

DmitriiTsy replied:
Sure. I should include more sense in naming functions/varibales or smth ellse

anatoliy8493 commented on src/components/Characters/Characters.js:
```diff
@@ -17,38 +36,175 @@ const Characters = () => {
+        </div>
+    ))
+
+    const HandlerFilter = (event) => {
```
the same applies everywhere

DmitriiTsy replied:
sure

anatoliy8493 commented on src/components/Characters/Characters.js:
```diff
@@ -17,38 +36,175 @@ const Characters = () => {
+                {error && <div>No characters found</div>}
+                <div className='filter-elements-wrapper'>
+                    {filter && 
+                        <div className='filter-wrapper'>
```
can you attach some screenshots to the PR to see how it looks like

DmitriiTsy commented on src/components/Characters/Characters.js:
```diff
@@ -17,38 +36,175 @@ const Characters = () => {
+                                    <button onClick={HandlerCurrentSpecies} value='Poopybutthole' style={currentSpecies === 'Poopybutthole' ? {backgroundColor: 'grey', color:'white'}:{backgroundColor: 'white', color:'grey'}}  className="filter-buttons-wrapper__element">Poopybutthole</button>
+                                    <button onClick={HandlerCurrentSpecies} value='Mythological' style={currentSpecies === 'Mythological' ? {backgroundColor: 'grey', color:'white'}:{backgroundColor: 'white', color:'grey'}}  className="filter-buttons-wrapper__element">Mythological</button>
+                                    <button onClick={HandlerCurrentSpecies} value='Unknown' style={currentSpecies === 'Unknown' ? {backgroundColor: 'grey', color:'white'}:{backgroundColor: 'white', color:'grey'}}  className="filter-buttons-wrapper__element">Unknown</button>
+                                    <button onClick={HandlerCurrentSpecies} value='Animal' style={currentSpecies === 'Animal' ? {backgroundColor: 'grey', color:'white'}:{backgroundColor: 'white', color:'grey'}}  className="filter-buttons-wrapper__element">Animal</button>
```
Fixed

DmitriiTsy commented:
<img width="938" alt="Screen Shot 2022-10-17 at 18 29 14" src="https://user-images.githubusercontent.com/101737790/196313980-9f95abb8-a65d-453e-895e-ac613657957b.png">

Screenshot when the filter is active and an error occurs

DmitriiTsy commented:
<img width="1402" alt="Screen Shot 2022-10-17 at 18 30 25" src="https://user-images.githubusercontent.com/101737790/196314138-de919a0e-dbd5-42e0-ae0a-924da196b65b.png">
When the filter is not active

DmitriiTsy committed:
characters-fixes

DmitriiTsy commented:
@anatoliy8493 Hi, could you check it again, please? Look's like I fixed everything you mentioned and added some filter/error screenshots to this PR. So, please let me know if missed smth

DmitriiTsy commented on src/index.css:
```diff
@@ -1,13 +0,0 @@
-body {
-  margin: 0;
```
That was my first commit to this PR. So I merged files to the main and then fetched them to a local folder. That's why you can't see what was inside. That was my mistake with deleting files. 

Maybe I can create another PR with all files that I have here, what do you think?

anatoliy8493 commented on src/components/Characters/Characters.js:
```diff
@@ -1,54 +1,147 @@
+  }, [info, url, search, results]);
+
+  useEffect(() => {
+    const controller = new AbortController(); // combined url + result in input for
```
make sure it works, use slow 3G simulation in Chrome dev tools

DmitriiTsy replied:
I checked 2 versions. One with an Abort controller and one without. So performance is the same as well as lack of errors. There are no errors in both cases.

anatoliy8493 commented on src/components/Filter/Filter.js:
```diff
@@ -0,0 +1,107 @@
+import React, { useState, useEffect } from 'react';
+
+function Filter(props) {
+  const [status, setStatus] = useState(false); // States to open additional filters
```
create hooks folder and use useReducer

DmitriiTsy replied:
Sure

DmitriiTsy replied:
Could you check please if there is more space to use Custom Hooks comparing what I chose with "use filter"?

anatoliy8493 commented on src/components/Characters/Character.css:
```diff
@@ -1,31 +1,102 @@
 
-.Character__image{
-    border-radius: 16px;
+.character__image{
```
space is missing, ensure ESLint works everywhere

DmitriiTsy replied:
done

anatoliy8493 commented on .eslintrc.json:
```diff
@@ -0,0 +1,21 @@
+        "commonjs": true,
+        "es2021": true
+    },
+    "extends": [
```
add prettier plugin

DmitriiTsy replied:
done

anatoliy8493 replied:
I see only react plugin ^

DmitriiTsy committed:
styling-css-fixes

DmitriiTsy committed:
added-one-reducer

DmitriiTsy commented on src/components/normalize/normalize.css:
```diff
@@ -0,0 +1,349 @@
+/*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */
```
Done

DmitriiTsy commented on src/components/Characters/Characters.js:
```diff
@@ -1,54 +1,147 @@
+    console.log('url: ', url);
+    console.log('info: ', info);
+    console.log('results', results);
+    console.log('search: ', search);
```
Done

DmitriiTsy committed:
reducer-fixes

DmitriiTsy commented:
<img width="1414" alt="Screen Shot 2022-10-21 at 15 16 02" src="https://user-images.githubusercontent.com/101737790/197296471-7fd1d2fc-2d13-4b99-b977-c20dea1059d7.png">
Hi, @anatoliy8493 Here are the last fixes. Whenever you have some time please check it

P.S. You could also download all files and launch the whole project locally

anatoliy8493 commented on src/components/Characters/Characters.js:
```diff
@@ -1,54 +1,201 @@
+  } = state;
+  const [addition, setAddition] = useState('');
+
+  const getDataValue = (value) => {
```
which data? what value? let's name it less general

DmitriiTsy replied:
Sure. You 100% right.

anatoliy8493 commented on src/components/Filter/Filter.js:
```diff
@@ -0,0 +1,174 @@
+import React, { useReducer } from 'react';
+import useFilter from '../hooks/useFilter';
+
+const initialState = {
```
let's move it to the hooks folder and name `useFilter`

DmitriiTsy replied:
You mean to transfer filter.js file to useFilter as a custom hook in the hook folder, am I right?

anatoliy8493 commented on src/components/Filter/Filter.js:
```diff
@@ -0,0 +1,174 @@
+    dispatch({ type: ACTIONS.CURRENT_GENDER, payload: event.target.id });
+  };
+
+  const handlerClean = (event) => {
```
this should be a separate action

DmitriiTsy replied:
Right

anatoliy8493 commented on src/components/Filter/Filter.js:
```diff
@@ -0,0 +1,174 @@
+  };
+
+  const handlerCurrentStatus = (event) => {
+    event.preventDefault();
```
do we really need preventDefault every time? what is this method doing?

DmitriiTsy replied:
This method prevents potential reload of the page after a button click. So, it's better to use it only with buttons

DmitriiTsy replied:
fixed

DmitriiTsy committed:
custom-hook-fixes

DmitriiTsy commented:
@anatoliy8493 Hi!

So, yeah, I encapsulated the whole logic of Filter to "use filter" custom hook. I also added a "useLink" hook to handle link creation. 

Could you check this, please?

DmitriiTsy commented on src/components/hooks/useFilter.js:
```diff
@@ -0,0 +1,137 @@
+  };
+
+  return [
+    handlerCurrentStatus,
```
Replace with object

DmitriiTsy commented on src/components/hooks/Filter.js:
```diff
@@ -0,0 +1,90 @@
+/* eslint-disable no-console */
```
Move to component folder

DmitriiTsy commented on src/components/hooks/Filter.js:
```diff
@@ -0,0 +1,90 @@
+/* eslint-disable no-console */
+/* eslint-disable jsx-a11y/no-static-element-interactions */
```
Move to component folder

DmitriiTsy commented on src/components/hooks/useFilter.js:
```diff
@@ -0,0 +1,137 @@
+/* eslint-disable no-console */
```
Remove all actions except setFilter and resetFilter + updateUI filter, use Wayfair example for it

DmitriiTsy commented on src/components/hooks/useLink.js:
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable no-console */
```
Add to useFilter

DmitriiTsy replied:
name to GeneretaSearch Linjk or smth like that
